Implements Blockwise lora

UmerHA opened the pull request:
# What does this PR do?
Allows setting LoRA weights more granularly, up to per-transformer block.

Fixes #7231

Example usage:
```python
pipe = ... # create pipeline
pipe.load_lora_weights(..., adapter_name="my_adapter") 
scales = {
    "text_encoder": 0.5,
    "unet": {
        "down": 0.9,  # all transformers in the down-part will use scale 0.9
        # "mid"  # because "mid" is not given, all transformers in the mid part will use the default scale 1.0
        "up": {
            "block_0": 0.6,  # all 3 transformers in the 0th block in the up-part will use scale 0.6
            "block_1": [0.4, 0.8, 1.0],  # the 3 transformers in the 1st block in the up-part will use scales 0.4, 0.8 and 1.0 respectively
        }
    }
}
pipe.set_adapters("my_adapter", scales)
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Core library:
- General functionalities: @sayakpaul @yiyixuxu @DN6

@dain5832: As you're the author of the solved issue, I invite you to also test this.

asomoza commented:
this PR is really nice, I like to be able to have this kind of control over the LoRAs. Just by playing with it I found a way to break it though. If you do this:

```py
scales = {
    "text_encoder": 0.5,
}
pipeline.set_adapters(["test"], adapter_weights=[scales])
```

it throws:

```
unet.py", line 611, in _expand_lora_scales_dict
    if "mid" not in scales:
TypeError: argument of type 'NoneType' is not iterable
```

Also I would like to request if you can add to have control over the second text encoder in SDXL but if it's rare edge case, it doesn't matter, I can do it on my side after.

UmerHA commented:
> this PR is really nice


> Just by playing with it I found a way to break it though. If you do this:

fixed

> 
> Also I would like to request if you can add to have control over the second text encoder in SDXL but if it's rare edge case, it doesn't matter, I can do it on my side after.

sayakpaul commented on docs/source/en/using-diffusers/loading_adapters.md:
````diff
@@ -157,6 +157,22 @@ image
+```python
+pipe = ... # create pipeline
+pipe.load_lora_weights(..., adapter_name="my_adapter") 
+scales = {
+    "text_encoder": 0.5,
+    "text_encoder_2": 0.5,  # only usable if pipe has a 2nd text encoder
+    "down": 0.9,  # all transformers in the down-part will use scale 0.9
+    # "mid"  # because "mid" is not given, all transformers in the mid part will use the default scale 1.0
+    "up": {
+        "block_0": 0.6,  # all 3 transformers in the 0th block in the up-part will use scale 0.6
+        "block_1": [0.4, 0.8, 1.0],  # the 3 transformers in the 1st block in the up-part will use scales 0.4, 0.8 and 1.0 respectively
+    }
+}
````
I think it's better to allow something like:

```python
scales = {
    "text_encoder": 0.5,
    "text_encoder_2": 0.5,  # only usable if pipe has a 2nd text encoder
    "unet": {
        "down": 0.9,  # all transformers in the down-part will use scale 0.9
        # "mid"  # because "mid" is not given, all transformers in the mid part will use the default scale 1.0
        "up": {
            "block_0": 0.6,  # all 3 transformers in the 0th block in the up-part will use scale 0.6
            "block_1": [0.4, 0.8, 1.0],  # the 3 transformers in the 1st block in the up-part will use scales 0.4, 0.8 and 1.0 respectively
        }
    }
}
```

I don't understand why there is a restriction on `mid` when imposing the `scale`. Could you elaborate? 

Also, what if I wanted to specify the scale for the conv layers too.

sayakpaul commented on src/diffusers/loaders/lora.py:
```diff
@@ -1033,17 +1036,73 @@ def enable_lora_for_text_encoder(self, text_encoder: Optional["PreTrainedModel"]
+        adapter_names = [adapter_names] if isinstance(adapter_names, str) else adapter_names
+
+        number = (float, int)
+        has_2nd_text_encoder = hasattr(self, "text_encoder_2")
```
```suggestion
        has_second_text_encoder = hasattr(self, "text_encoder_2")
```

sayakpaul commented on src/diffusers/loaders/lora.py:
```diff
@@ -1033,17 +1036,73 @@ def enable_lora_for_text_encoder(self, text_encoder: Optional["PreTrainedModel"]
     ):
+        adapter_names = [adapter_names] if isinstance(adapter_names, str) else adapter_names
+
+        number = (float, int)
```
```suggestion
        allowed_numeric_dtypes = (float, int)
```

sayakpaul commented on src/diffusers/loaders/unet.py:
```diff
@@ -561,6 +561,100 @@ def _unfuse_lora_apply(self, module):
+
+            del scales[updown]
+
+        def layer_name(name):
```
Can keep it private as well?

UmerHA replied:
It's defined inside the function `_expand_lora_scales_dict`, so is not available publicly.

sayakpaul commented on src/diffusers/loaders/unet.py:
```diff
@@ -561,6 +561,100 @@ def _unfuse_lora_apply(self, module):
             if isinstance(module, BaseTunerLayer):
                 module.unmerge()
 
+    def _expand_lora_scales_dict(
```
Why not allow conv layers here as well?

sayakpaul commented on src/diffusers/loaders/unet.py:
```diff
@@ -561,6 +561,100 @@ def _unfuse_lora_apply(self, module):
+        self, scales, blocks_with_transformer: Dict[str, int], transformer_per_block: Dict[str, int]
+    ):
+        """
+        Expand input into a weight dict with a weight per transformer
```
```suggestion
        Expands the inputs into a more granular dictionary. See the example below for more details. 
```

sayakpaul commented on src/diffusers/loaders/unet.py:
```diff
@@ -599,14 +693,27 @@ def set_adapters(
 
+        # Set missing value to default of 1.0
+        weights = [weight or 1.0 for weight in weights]
+        blocks_with_transformer = {
+            "down": [i for i, block in enumerate(self.down_blocks) if hasattr(block, "attentions")],
+            "up": [i for i, block in enumerate(self.up_blocks) if hasattr(block, "attentions")],
+        }
```
Same questions:

* Why not allow the `mid` block? 
* Why not allow `conv`s?

sayakpaul reviewed (commented):
I think the PR has a very nice start. I left some initial comments. We'll need to make sure to add rigorous checks on the inputs to prevent anything unexpected that we can foresee from our experience.

UmerHA commented:
**Re `mid` block:** We can pass a "mid" key in the scales dict. It's commented out to show not all keys need to be provided. But the `mid` block only needs 1 scale number, because it is only 1 block and has only 1 layer.

In principal, `mid` blocks could have more layers, but `UNet2DConditionModel` doesn't pass `num_layers` to the `mid` block, so it always defaults to 1.

**Re convs:** The [original implementation](https://github.com/hako-mikan/sd-webui-lora-block-weight) doesn't implements scaling the conv layers.

---

- Would you also implement scaling the conv layers?
- Would you also implement scaling the resnets? Currently, only the transformer are scaled.

Let me know, then I'll adapt the code.

@sayakpaul

sayakpaul commented:
> Re mid block: We can pass a "mid" key in the scales dict. It's commented out to show not all keys need to be provided. But the mid block only needs 1 scale number, because it is only 1 block and has only 1 layer.

I still don't see why we shouldn't make it configurable as well actually. Even if it's just a single block there might be effects if the users can control its contributions. So, I would vote for the ability to configure it too. 

> Would you also implement scaling the conv layers?
> Would you also implement scaling the resnets? Currently, only the transformer are scaled.

All of them should configurable IMO if we were to support this feature otherwise the design might feel very convoluted. But I would like to also take opinions from @BenjaminBossan and @yiyixuxu here.

UmerHA commented:
> I still don't see why we shouldn't make it configurable as well actually. Even if it's just a single block there might be effects if the users can control its contributions. So, I would vote for the ability to configure it too.

I think we're miscommunicating. With the current PR, You CAN already configure the `mid` block:

```python
pipe = ... # create pipeline
pipe.load_lora_weights(..., adapter_name="my_adapter") 
scales = { "mid" : 0.5 }
pipe.set_adapters("my_adapter", scales)
```
However, you can only pass a single number to it, because it has in total only 1 transformer layer in it. Unlike the up / down part, which have multiple blocks and multiple layers per block. That's why you can pass a `Dict` (per block) or `Dict[List]`(per block, per layer) to it.

sayakpaul commented:
> However, you can only pass a single number to it, because it has in total only 1 transformer layer in it. Unlike the up / down part, which have multiple blocks and multiple layers per block. That's why you can pass a Dict (per block) or Dict[List](per block, per layer) for them.

Ah makes sense then.

BenjaminBossan commented on src/diffusers/loaders/lora.py:
```diff
@@ -986,6 +986,9 @@ def process_weights(adapter_names, weights):
                     f"Length of adapter names {len(adapter_names)} is not equal to the length of the weights {len(weights)}"
                 )
+
+            weights = [{"text_model": w} if w is not None else {"text_model": 1.0} for w in weights]
```
Technically, `w` cannot be `None` here, right? This would require the weights to be a list with None elements.

BenjaminBossan commented on src/diffusers/loaders/lora.py:
```diff
@@ -1033,17 +1036,73 @@ def enable_lora_for_text_encoder(self, text_encoder: Optional["PreTrainedModel"]
+        has_2nd_text_encoder = hasattr(self, "text_encoder_2")
+
+        # Expand weights into a list, one entry per adapter
+        if adapter_weights is None or isinstance(adapter_weights, (number, dict)):
```
I didn't know that `isinstance` accepts nested tuples as 2nd argument :-O

UmerHA replied:
It does! :)

BenjaminBossan commented on src/diffusers/loaders/lora.py:
```diff
@@ -1033,17 +1036,73 @@ def enable_lora_for_text_encoder(self, text_encoder: Optional["PreTrainedModel"]
+            elif weight is None:
+                return {key: 1.0 for key in allowed_keys}
+            else:
+                raise RuntimeError(f"lora weight has wrong type {type(weight)}.")
```
Overall, I find the usage of the word "weight" in this context confusing, because the parameters are also often called "weight", whereas here we mean the scales, right? I know this wording was already present before this PR, but I still wonder if we should be more precise in the error messages, e.g. use "weight scale" or something like that instead of only "weight". WDYT?

BenjaminBossan commented on src/diffusers/loaders/lora.py:
```diff
@@ -1033,17 +1036,73 @@ def enable_lora_for_text_encoder(self, text_encoder: Optional["PreTrainedModel"]
+            else:
+                raise RuntimeError(f"lora weight has wrong type {type(weight)}.")
+
+        adapter_weights = [ensure_is_dict(weight) for weight in adapter_weights]
```
To understand the flow of the code, it would be really helpful to have a type annotation or a comment here, as the type of `adapter_weights` changes all the time.

BenjaminBossan commented on src/diffusers/loaders/lora.py:
```diff
@@ -1033,17 +1036,73 @@ def enable_lora_for_text_encoder(self, text_encoder: Optional["PreTrainedModel"]
         self,
         adapter_names: Union[List[str], str],
-        adapter_weights: Optional[List[float]] = None,
+        adapter_weights: Optional[Union[List[float], List[Dict]]] = None,
```
Can the type annotation be more precise than `Dict`? I think it's `Dict[str, Union[float, int]]`?

BenjaminBossan commented on src/diffusers/loaders/lora.py:
```diff
@@ -1033,17 +1036,73 @@ def enable_lora_for_text_encoder(self, text_encoder: Optional["PreTrainedModel"]
+            self.set_adapters_for_text_encoder(adapter_names, self.text_encoder, text_encoder_weights)
         if hasattr(self, "text_encoder_2"):
-            self.set_adapters_for_text_encoder(adapter_names, self.text_encoder_2, adapter_weights)
+            self.set_adapters_for_text_encoder(adapter_names, self.text_encoder_2, text_encoder_2_weights)
```
Line above could be replaced with `if has_2nd_text_encoder:`

BenjaminBossan commented on src/diffusers/loaders/lora.py:
```diff
@@ -1033,17 +1036,73 @@ def enable_lora_for_text_encoder(self, text_encoder: Optional["PreTrainedModel"]
+                if k not in allowed_keys:
+                    raise ValueError(
+                        f"Got invalid key '{k}' in lora weight dict. Allowed keys are 'text_encoder', 'text_encoder_2', 'down', 'mid', 'up'."
+                    )
```
I wonder if it would help readability if all the new code above was factored out into a utility function like `def normalize_adapter_weight_argument(adapter_weights, has_2nd_text_encoder)` or so. Maybe diffusers maintainers have an opinion on that. Such a function would also be easier to unit-test.

BenjaminBossan commented on src/diffusers/loaders/unet.py:
```diff
@@ -561,6 +561,100 @@ def _unfuse_lora_apply(self, module):
+            transformer_per_block (`Dict[str, int]`):
+                Dict with keys 'up' and 'down', showing how many transformer layers each block has
+
+        E.g. turns
```
I think the example would be more informative if it showed this method being called with certain arguments and the corresponding output.

UmerHA replied:
Agree - have expanded the example

BenjaminBossan commented on src/diffusers/loaders/unet.py:
```diff
@@ -561,6 +561,100 @@ def _unfuse_lora_apply(self, module):
+        Expand input into a weight dict with a weight per transformer
+
+        Parameters:
+            blocks_with_transformer (`Dict[str, int]`):
```
`scales` argument is missing.

BenjaminBossan commented on src/diffusers/loaders/unet.py:
```diff
@@ -561,6 +561,100 @@ def _unfuse_lora_apply(self, module):
+        for layer in scales.keys():
+            if not any(layer_name(layer) in module for module in state_dict.keys()):
+                raise ValueError(
+                    f"Can't set lora scale for layer {layer}. It either doesn't exist in this unet or has not attentions."
```
```suggestion
                    f"Can't set lora scale for layer {layer}. It either doesn't exist in this unet or it is not an attention layer."
```
?

BenjaminBossan reviewed (commented):
Thanks for adding this feature. I don't have sufficient knowledge to judge if this covers all edge cases, so I'll leave that to the experts. Still added a few comments.

stevhliu commented on docs/source/en/using-diffusers/loading_adapters.md:
````diff
@@ -157,6 +157,22 @@ image
 
 For both [`~loaders.LoraLoaderMixin.load_lora_weights`] and [`~loaders.UNet2DConditionLoadersMixin.load_attn_procs`], you can pass the `cross_attention_kwargs={"scale": 0.5}` parameter to adjust how much of the LoRA weights to use. A value of `0` is the same as only using the base model weights, and a value of `1` is equivalent to using the fully finetuned LoRA.
 
+For fine-grained control on how much of the LoRA weights are used, use [`~loaders.LoraLoaderMixin.set_adapters`]. Here, you can define scale of any granularity up to per-transformer.
+```python
+pipe = ... # create pipeline
+pipe.load_lora_weights(..., adapter_name="my_adapter") 
+scales = {
+    "text_encoder": 0.5,
+    "text_encoder_2": 0.5,  # only usable if pipe has a 2nd text encoder
+    "down": 0.9,  # all transformers in the down-part will use scale 0.9
+    # "mid"  # because "mid" is not given, all transformers in the mid part will use the default scale 1.0
+    "up": {
+        "block_0": 0.6,  # all 3 transformers in the 0th block in the up-part will use scale 0.6
+        "block_1": [0.4, 0.8, 1.0],  # the 3 transformers in the 1st block in the up-part will use scales 0.4, 0.8 and 1.0 respectively
+    }
+}
+pipe.load_lora_weights("my_adapter", scales)
+```
````
Since this is more substantial now, I think it'd be better to create a new subsection. Something like:

```md
### Adjust LoRA weight scale

For both load_lora_weights() and load_attn_procs()...

new content about using `scales`
```

stevhliu commented on docs/source/en/using-diffusers/loading_adapters.md:
```diff
@@ -157,6 +157,22 @@ image
 
 For both [`~loaders.LoraLoaderMixin.load_lora_weights`] and [`~loaders.UNet2DConditionLoadersMixin.load_attn_procs`], you can pass the `cross_attention_kwargs={"scale": 0.5}` parameter to adjust how much of the LoRA weights to use. A value of `0` is the same as only using the base model weights, and a value of `1` is equivalent to using the fully finetuned LoRA.
 
+For fine-grained control on how much of the LoRA weights are used, use [`~loaders.LoraLoaderMixin.set_adapters`]. Here, you can define scale of any granularity up to per-transformer.
```
```suggestion
For more granular control over the amount of LoRA weights used per layer, you can use `scales` to define a dictionary specifying how much to scale the weights in each layer by.
```

UmerHA replied:
The example had an error. You use `pipe.set_adapters` (not `pipe.load_lora_weights`) to pass the `scales` dictionary.

Is this okay:
> For more granular control on the amount of LoRA weights used per layer, you can use [`~loaders.LoraLoaderMixin.set_adapters`] and pass a `scales` dictionary specifying how much to scale the weights in each layer by.

?

UmerHA replied:
My example had an error. You use `pipe.set_adapters` (not `pipe.load_lora_weights`) to pass the dictionary, and the dictionary param is called `adapter_weights`.

This is unfortunately a bit confusing, because `weights` also means the actual lora weights, and also the scale weights. But the naming was already here before this PR.

Does this work for you:

> For more granular control on the amount of LoRA weights used per layer, you can use [`~loaders.LoraLoaderMixin.set_adapters`] and pass a dictionary specifying how much to scale the weights in each layer by.

?

stevhliu replied:
Yeah that'll work, thanks for clarifying :)

stevhliu reviewed (approved):
Awesome, great work! 👏

yiyixuxu commented on src/diffusers/loaders/lora.py:
```diff
@@ -1033,17 +1036,71 @@ def enable_lora_for_text_encoder(self, text_encoder: Optional["PreTrainedModel"]
+        if has_second_text_encoder:
+            text_encoder_2_weights = []
+
+        for weights in adapter_weights:
+            unet_weight = weights.get("unet", None)
+            text_encoder_weight = weights.get("text_encoder", None)
+            if has_second_text_encoder:
+                text_encoder_2_weight = weights.get("text_encoder_2", None)
+
+            unet_weights.append(unet_weight)
+            text_encoder_weights.append(text_encoder_weight)
+            if has_second_text_encoder:
+                text_encoder_2_weights.append(text_encoder_2_weight)
```
```suggestion
        for adapter_name, weights in zip(adapter_names, adapter_weights):
            if isinstance(weights, dict):
                unet_weight = weights.pop("unet", None)]
                text_encoder_weight = weights.pop("text_encoder", None)
                text_encoder_2_weight = weights.pop("text_encoder_2", None)
                if len(weights) >0:
                    raise ValueError(f"Got invalid key '{weights.keys()}' in lora weight dict for adapter {adapter_name}.")
               if text_encoder_2_weight is not None and not hasattr(self, "text_encoder_2"):
                   logger.warning("lora weight dict contains text_encoder_2 weights but will be ignored because pipeline does not have text_encoder_2")
            else:
               unet_weight = weights
               text_encoder_weight = weights
               text_encoder_2_weight = weights
       
            unet_weights.append(unet_weight)
            text_encoder_weights.append(text_encoder_weight)
            text_encoder_2_weights.append(text_encoder_2_weight)
```

yiyixuxu commented on src/diffusers/loaders/unet.py:
```diff
@@ -561,10 +561,104 @@ def _unfuse_lora_apply(self, module):
+                'down': 2,
+                'mid': 3,
+                'up': {
+                    'block_1': 4,
+                    'block_2': [5, 6, 7]
```
did I understand it wrong or it is a typo?
```suggestion
                    'block_0': 4,
                    'block_1': [5, 6, 7]
```

UmerHA replied:
It's a typo. Have fixed it.

yiyixuxu reviewed (commented):
hey! thanks for your PR
I did first round of review - let me know if it makes sense

Thanks
YiYi

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_7352). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

BenjaminBossan commented on src/diffusers/loaders/unet.py:
```diff
@@ -597,16 +700,27 @@ def set_adapters(
                 f"Length of adapter names {len(adapter_names)} is not equal to the length of their weights {len(weights)}."
             )
 
+        weights = [weight or 1.0 for weight in weights]  # Set None values to default of 1.0
```
same comment as earlier

BenjaminBossan reviewed (commented):
LGTM overall, nice tests. I just have some smaller comments, please check.

All in all, I don't know enough about diffusers to give a final approval, so I'll leave that to the experts.

UmerHA commented:
@sayakpaul

> > Would you also implement scaling the conv layers?
> > Would you also implement scaling the resnets? Currently, only the transformer are scaled.
> 
> All of them should configurable IMO if we were to support this feature otherwise the design might feel very convoluted. 

To make all of the modules configurable, I propose the dictionary structure below. The example is for SDXL and shows the full dictionary. Of course, the user can pass only a partial dict or just a number, and that will be expanded like in the current PR.

Does that work for you and @yiyixuxu? If yes, then I'd adapt the PR accordingly.

```python
{
    "text_encoder": 1.0,
    "text_encoder_2": 1.0,
    "unet": {
        "down": {
            "block_0": {
                "resnet_0": 1.0,
                "resnet_1": 1.0,
                "downsampler": 1.0,
            },
            "block_1": {
                "resnet_0": 1.0,
                "attention_0": 1.0,
                "resnet_1": 1.0,
                "attention_1": 1.0,
                "downsampler": 1.0,
            },
            "block_2": {
                "resnet_0": 1.0,
                "attention_0": 1.0,
                "resnet_1": 1.0,
                "attention_1": 1.0
            }
        },
        "mid":  {
            "resnet_0": 1.0,
            "attention_0": 1.0,
            "resnet_1": 1.0
        }
        "up": {
            "block_0": {
                "resnet_0": 1.0,
                "attention_0": 1.0,
                "resnet_1": 1.0,
                "attention_1": 1.0,
                "resnet_2": 1.0,
                "attention_2": 1.0,
                "upsampler": 1.0
            },
            "block_1": {
                "resnet_0": 1.0,
                "attention_0": 1.0,
                "resnet_1": 1.0,
                "attention_1": 1.0,
                "resnet_2": 1.0,
                "attention_2": 1.0,
                "upsampler": 1.0
            },
            "block_2": {
                "resnet_0": 1.0,
                "resnet_1": 1.0,
                "resnet_2": 1.0,
            }
        }
    }
}
```

yiyixuxu commented:
@UmerHA
I'm ok with it - 
@sayakpaul what do you think? should we also add a property something like `default_lora_weight` for the user to get this map and update it?

sayakpaul commented on docs/source/en/using-diffusers/loading_adapters.md:
```diff
@@ -153,18 +153,35 @@ image
+        }
+    }
+}
+pipe.set_adapters("my_adapter", scales)
```
How would this pan out for multiple adapters?

sayakpaul commented on docs/source/en/using-diffusers/loading_adapters.md:
```diff
@@ -153,18 +153,35 @@ image
+        }
+    }
+}
+pipe.set_adapters("my_adapter", scales)
```
How would this pan out for multiple adapters?

UmerHA replied:
`pipe.set_adapters(["my_adapter_1", "my_adapter_2"], [scales_1, scales_2])`

will add a remark.

sayakpaul commented on src/diffusers/loaders/lora.py:
```diff
@@ -959,7 +960,7 @@ def set_adapters_for_text_encoder(
         adapter_names: Union[List[str], str],
         text_encoder: Optional["PreTrainedModel"] = None,  # noqa: F821
-        text_encoder_weights: List[float] = None,
+        text_encoder_weights: Optional[Union[float, List[float], List[NoneType]]] = None,
```
What does `List[NoneType]` signify? What kind of inputs are expected there?

UmerHA replied:
This function can receive a list of `None`s. Then for each adapter the default value of `1.0` is taken.

I've changes this because `LoraLoaderMixin.set_adapters_for_text_encoder` calls this function and can now pass a list of Nones.

We need to either add the type hint `List[NoneType]` here or add
```
if all(w is None for w in text_encoder_weights):
    text_encoder_weights = None
```
in `LoraLoaderMixin.set_adapters_for_text_encoder`

sayakpaul replied:
Thanks for explaining!

sayakpaul commented on src/diffusers/loaders/lora.py:
```diff
@@ -977,15 +978,17 @@ def set_adapters_for_text_encoder(
-            if weights is None:
-                weights = [1.0] * len(adapter_names)
-            elif isinstance(weights, float):
-                weights = [weights]
+            if not isinstance(weights, list):
+                weights = [weights] * len(adapter_names)
```
We should not be deviating from the default values here. These changes remove the default value of 1.0 from `weights`. 

Edit: I see you're tackling it a little later. Okay for me.

sayakpaul commented on src/diffusers/loaders/lora.py:
```diff
@@ -977,15 +978,17 @@ def set_adapters_for_text_encoder(
                 )
+
+            weights = [w or 1.0 for w in weights]  # Set None values to default of 1.0
+            weights = [{"text_model": w} for w in weights]
```
(nit): Providing some example outputs of these intermediate variables during their expansion would be helpful for the reviewers as well as the maintainers. WDYT?

UmerHA replied:
sure, added :)

sayakpaul commented on src/diffusers/loaders/lora.py:
```diff
@@ -1033,17 +1036,55 @@ def enable_lora_for_text_encoder(self, text_encoder: Optional["PreTrainedModel"]
+                f"Length of adapter names {len(adapter_names)} is not equal to the length of the weights {len(adapter_weights)}"
+            )
+
+        # Decompose weights into weights for unet, text_encoder and text_encoder_2
+        unet_weights, text_encoder_weights, text_encoder_2_weights = [], [], []
```
Stable Diffusion family of LoRAs won't have text encoder 2. How are we handling that? 

I think it should not be only applicable to SDXL LoRA, right?

sayakpaul replied:
Also, could we maybe update the variable names to `unet_lora_weights`, and so on to make things a little clearer?

sayakpaul commented on src/diffusers/loaders/lora.py:
```diff
@@ -1033,17 +1036,55 @@ def enable_lora_for_text_encoder(self, text_encoder: Optional["PreTrainedModel"]
+                        f"Got invalid key '{weights.keys()}' in lora weight dict for adapter {adapter_name}."
+                    )
+
+                if text_encoder_2_weight is not None and not hasattr(self, "text_encoder_2"):
+                    logger.warning(
+                        "Lora weight dict contains text_encoder_2 weights but will be ignored because pipeline does not have text_encoder_2."
+                    )
```
Ah nice! I think this addresses my previous comment.

sayakpaul commented on src/diffusers/loaders/unet.py:
```diff
@@ -561,10 +563,117 @@ def _unfuse_lora_apply(self, module):
+                Dict with keys 'up' and 'down', showing how many transformer layers each block has
+
+        E.g. turns
+            scales = {
+                'down': 2,
+                'mid': 3,
+                'up': {
+                    'block_0': 4,
+                    'block_1': [5, 6, 7]
+                }
+            }
+            blocks_with_transformer = {
+                'down': [1,2],
+                'up': [0,1]
+            }
+            transformer_per_block = {
+                'down': 2,
+                'up': 3
+            }
```
Let's maybe wrap this into markdown so that it renders appropriately.

sayakpaul commented on src/diffusers/loaders/unet.py:
```diff
@@ -561,10 +563,117 @@ def _unfuse_lora_apply(self, module):
+                'up': 3
+            }
+        into
+            {
+                'down.block_1.0': 2,
+                'down.block_1.1': 2,
+                'down.block_2.0': 2,
+                'down.block_2.1': 2,
+                'mid': 3,
+                'up.block_0.0': 4,
+                'up.block_0.1': 4,
+                'up.block_0.2': 4,
+                'up.block_1.0': 5,
+                'up.block_1.1': 6,
+                'up.block_1.2': 7,
+            }
```
Same.

sayakpaul commented on tests/lora/test_lora_layers_peft.py:
```diff
@@ -817,6 +818,226 @@ def test_simple_inference_with_text_unet_multi_adapter(self):
+                    self.check_if_lora_correctly_set(pipe.text_encoder_2), "Lora not correctly set in text encoder 2"
+                )
+
+            weights_1 = {
+                "unet": {
+                    "down": 5,
+                }
+            }
```
Can we collapse this dictionary into a single line please?

sayakpaul commented on tests/lora/test_lora_layers_peft.py:
```diff
@@ -817,6 +818,226 @@ def test_simple_inference_with_text_unet_multi_adapter(self):
+            pipe.set_adapters("adapter-1", weights_1)
+            output_weights_1 = pipe(**inputs, generator=torch.manual_seed(0)).images
+
+            weights_2 = {
+                "unet": {
+                    "up": 5,
+                }
+            }
```
Same.

sayakpaul commented on tests/lora/test_lora_layers_peft.py:
```diff
@@ -817,6 +818,226 @@ def test_simple_inference_with_text_unet_multi_adapter(self):
                 "output with no lora and output with lora disabled should give same results",
             )
 
+    def test_simple_inference_with_text_unet_block_scale(self):
```
Since the test has `text` inside of it, let's also target some text encoder layers in the `scales`.

sayakpaul commented on tests/lora/test_lora_layers_peft.py:
```diff
@@ -817,6 +818,226 @@ def test_simple_inference_with_text_unet_multi_adapter(self):
+                "output with no lora and output with lora disabled should give same results",
+            )
+
+    def test_simple_inference_with_text_unet_multi_adapter_block_lora(self):
```
Same. Let's target `text_encoder` layers in the scales.

sayakpaul commented on tests/lora/test_lora_layers_peft.py:
```diff
@@ -817,6 +818,226 @@ def test_simple_inference_with_text_unet_multi_adapter(self):
+
+            return opts
+
+        components, text_lora_config, unet_lora_config = self.get_dummy_components(self.scheduler_cls)
+        pipe = self.pipeline_class(**components)
+        pipe = pipe.to(self.torch_device)
+        pipe.set_progress_bar_config(disable=None)
+        _, _, inputs = self.get_dummy_inputs(with_generator=False)
+
+        pipe.text_encoder.add_adapter(text_lora_config, "adapter-1")
+        pipe.unet.add_adapter(unet_lora_config, "adapter-1")
+
+        if self.has_two_text_encoders:
+            pipe.text_encoder_2.add_adapter(text_lora_config, "adapter-1")
+
+        for scale_dict in all_possible_dict_opts(pipe.unet, value=1234):
+            # test if lora block scales can be set with this scale_dict
+            if not self.has_two_text_encoders and "text_encoder_2" in scale_dict:
+                del scale_dict["text_encoder_2"]
+
+            pipe.set_adapters("adapter-1", scale_dict)
```
What is this testing? I don't see any assertions here. Is this expected?

UmerHA replied:
This tests that all possible input dicts can be passed into `pipe.set_adapters` and the dict expansion works.

If it doesn't, the line `pipe.set_adapters("adapter-1", scale_dict)` will throw an error and the test fails.
I'll add a comment:
```python
pipe.set_adapters("adapter-1", scale_dict)  # test will fail if this line throws an error
```

sayakpaul reviewed (commented):
Thanks for working on this! Very impressive PR. 

I have left a couple of questions and comments. Let's make sure we set the expectations right for our users by mentioning that this features doesn't currently target the convolution modules and they will be defaulted to a scale of `1.0` when using this feature. Did I get that correct?

UmerHA commented:
@sayakpaul thanks for the thorough review. 

You [had said](https://github.com/huggingface/diffusers/pull/7352#issuecomment-2003564430) all modules should be configurable. For that I proposed a dict structure in [this comment above](https://github.com/huggingface/diffusers/pull/7352#issuecomment-2007205049).

Would you rather I
1) Finish the PR as it currently is, and make it clear in the docs that only attentions will be scaled, or 
2) Adapt the PR as suggested in the comment and make every module configurable?

I'd vote for 2. I think you're right that it's useful and then it's done now.

Let me know & I'll get this done asap. ;)

sayakpaul commented:
To keep the scope of the PR manageable, I think it's okay to keep some blocks non-configurable and let the community play with it. Based on the community interest, we can take it forward. @yiyixuxu @BenjaminBossan WDYT?

sayakpaul commented on docs/source/en/using-diffusers/loading_adapters.md:
````diff
@@ -153,18 +153,45 @@ image
 pipeline.unload_lora_weights()
 ```
 
+### Adjust LoRA weight scale
````
@stevhliu looks good to you as far as the doc-related changes are concerned?

stevhliu replied:
Yeah docs look good on my end! 👍

sayakpaul reviewed (commented):
This is looking really solid. Thanks so much for working on this. 

Will provide my approval after other maintainers have had a chance to look through this.

BenjaminBossan reviewed (approved):
From what I can judge, this looks good, thanks a lot.

UmerHA commented on docs/source/en/using-diffusers/loading_adapters.md:
```diff
@@ -153,18 +153,45 @@ image
+pipe.load_lora_weights(..., adapter_name="my_adapter_2") 
+scales_1 = { ... }
+scales_2 = { ... }
+pipe.set_adapters(["my_adapter_1", "my_adapter_2"], [scales_1, scales_2])
```
We have a unit test for that in `test_simple_inference_with_text_unet_multi_adapter_block_lora`

I've expanded it to test an error is raised for mismatching element lengths:
```py
# a mismatching number of adapter_names and adapter_weights should raise an error
with self.assertRaises(ValueError):
    pipe.set_adapters(["adapter-1", "adapter-2"], [scales_1])
```

UmerHA commented:
> @UmerHA really impressive work! Would be curious to check how results with real LoRAs get affected with this feature. If you could provide amend some of the examples [here](https://huggingface.co/docs/diffusers/main/en/tutorials/using_peft_for_inference) with this feature and show us some results, that would be great!

@sayakpaul I've expanded that tutorial with a _Customize adapters strength_ section and added real examples. Does that work for you?

Here's a comparison of the images:

![all](https://github.com/huggingface/diffusers/assets/40663591/4b81b36d-986d-4ffa-a68d-74993991135c)

If that works for you, I'll send you the image via mail, so you can upload them.

sayakpaul commented:
Thanks. However, I don't think the pictures are reflecting the LoRA's effect at all. The "pixel" LoRA should be used with the trigger "pixel art".

UmerHA commented:
> Thanks. However, I don't think the pictures are reflecting the LoRA's effect at all. The "pixel" LoRA should be used with the trigger "pixel art".

Thanks for explaining. Here are the images for the prompt "toy_face of a hacker with a hoodie, pixel art"

![all](https://github.com/huggingface/diffusers/assets/40663591/e0ed5034-8910-41f1-9470-fd61fffd5663)

sayakpaul commented:
Hmm, so the LoRAs you're using here don't have text encoder counterparts. So, the effect is still not appropriate. We should be using the `unet` here. This leads me to think if we should warn the users about this situation i.e., making them aware that they have passed `scales` for modules whose counter LoRA parts aren't present in the first place.

yiyixuxu commented on src/diffusers/loaders/unet.py:
```diff
@@ -561,10 +562,121 @@ def _unfuse_lora_apply(self, module):
             if isinstance(module, BaseTunerLayer):
                 module.unmerge()
 
+    def _expand_lora_scales_dict(
```
should we make a `unet_loader_utils` file? move all these added functions there? cc @sayakpaul

sayakpaul replied:
Works for me.

UmerHA replied:
@yiyixuxu There's only 1 added function (`_expand_lora_scales_dict`) in `loaders/unet.py`. Are you sure we should create a new file for that?

Or are there other functions you're referring to?

yiyixuxu replied:
see this comment here :) https://github.com/huggingface/diffusers/pull/7352/files#r1535961194

sayakpaul commented on docs/source/en/tutorials/using_peft_for_inference.md:
```diff
@@ -133,6 +133,73 @@ image
 ![no-lora](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/peft_integration/diffusers_peft_lora_inference_20_1.png)
 
+### Customize adapters strength
+For even more customization, you can control how strongly the adapter affects each part of the pipeline. For this, pass a dictionary with the control strengths (called "scales") to [`~diffusers.loaders.UNet2DConditionLoadersMixin.set_adapters`].
```
> It's still unclear to me how this feature would allow me to control the output. I know I would be able to adjust the strength of Lora on the block level, but how would it correspond to the change in output? Are there any general rules you can follow?

You're basically imposing a granular level of control as to how you want to apply the adapter weights corresponding to the base layers. Think of applying a single adapter `scale` but applied on a more granular level. So, we control how much of the "adapter-effect" we want to see in the final generated output.

asomoza replied:
Sometimes people come with some theories about how it affects the result but none of those are true in my tests. I use this kind of precision for when I use multiple LoRAs and when I want to use one that is too strong or even with overfitted ones.

The variation in the result depends on the prompt so the only way of knowing this is just testing, for this, people in automatic1111 use the xyz plot to know which combination works best.

In the sd-webui extension they show this image: https://user-images.githubusercontent.com/55250869/270830887-dff65f45-823a-4dbd-94c5-34d37c84a84f.jpg and as you can see there's no bad generation but just different variations.

Once this PR is merged, I plan to do more extensive testing and post what I learn.

UmerHA commented:
> Hmm, so the LoRAs you're using here don't have text encoder counterparts. So, the effect is still not appropriate.

Okay, have used [this lora (ps1)](veryVANYA/ps1-graphics-sdxl-v2) which contains a `text_encoder`

![all](https://github.com/huggingface/diffusers/assets/40663591/eff6465d-ce56-419d-810b-6a64c29b3eb9)

> This leads me to think if we should warn the users about this situation i.e., making them aware that they have passed `scales` for modules whose counter LoRA parts aren't present in the first place.

Have added a warning

sayakpaul commented:
> Okay, have used [this lora (ps1)](https://github.com/huggingface/diffusers/pull/veryVANYA/ps1-graphics-sdxl-v2) which contains a text_encoder

The link is invalid. The associated prompt also needs to ensure that it has the trigger word. Can we just a LoRA from the guide I mentioned above and show a simple comparative study given we already know that it doesn't have a text encoder?

yiyixuxu commented on src/diffusers/loaders/unet.py:
```diff
@@ -597,16 +709,32 @@ def set_adapters(
+        # e.g. [7,7] -> [7,7] ; [None, None] -> [1.0, 1.0]
+        weights = [w if w is not None else 1.0 for w in weights]
+
+        blocks_with_transformer = {
+            "down": [i for i, block in enumerate(self.down_blocks) if hasattr(block, "attentions")],
+            "up": [i for i, block in enumerate(self.up_blocks) if hasattr(block, "attentions")],
+        }
+        transformer_per_block = {"down": self.config.layers_per_block, "up": self.config.layers_per_block + 1}
+
+        # e.g. [7,7] -> [{...}, {...}]
+        weights = [
+            self._expand_lora_scales_dict(weight_for_adapter, blocks_with_transformer, transformer_per_block)
+            for weight_for_adapter in weights
+        ]
+
```
can we put the entire loop into a function and move it to a separate util file along with `_expand_lora_scales_dict`? this mean the logic of `blocks_with_transformers` and transformer_per_block` will also be moved there

```suggestion 
        weights = ._expand_lora_scales_dicts(weights)
```

yiyixuxu replied:
for
```python
weights = _expand_lora_scales_dicts(self, weights)
```
also let's make sure that if `weights` is a not dict, do nothing

yiyixuxu commented on src/diffusers/loaders/unet.py:
```diff
@@ -561,10 +562,121 @@ def _unfuse_lora_apply(self, module):
+
+        scales = copy.deepcopy(scales)
+
+        if not isinstance(scales, dict):
+            scales = {o: scales for o in ["down", "mid", "up"]}
```
let's only expand it if it is a `dict`
```suggestion
```

yiyixuxu commented on src/diffusers/utils/peft_utils.py:
```diff
@@ -228,16 +228,22 @@ def delete_adapter_layers(model, adapter_name):
     from peft.tuners.tuners_utils import BaseTunerLayer
 
+    def get_module_weight(weight_for_adapter, module_name):
+        for layer_name, weight_ in weight_for_adapter.items():
+            if layer_name in module_name:
+                return weight_
+        raise RuntimeError(f"No LoRA weight found for module {module_name}, which should never happen.")
```
```suggestion
        if not isinstance(weight_for_adapter, dict):
            return weight_for_adapter
        for layer_name, weight_ in weight_for_adapter.items():
            if layer_name in module_name:
                return weight_
        raise ValueError(f"No LoRA weight found for module {module_name}")
```

yiyixuxu commented on src/diffusers/loaders/lora.py:
```diff
@@ -977,15 +978,23 @@ def set_adapters_for_text_encoder(
+            weights = [w if w is not None else 1.0 for w in weights]
+
+            # e.g. [3,1] -> [{"text_model": 3} , {"text_model": 1} ]
+            weights = [{"text_model": w} for w in weights]
```
```suggestion
```

yiyixuxu reviewed (commented):
I left a few more feedback, basically, I think `set_weights_and_activate_adapters` should still be able to handle the `weights` argument in the previous format (a list of scalar or float); this way we only need to expand the `weights` when it is a `dict` and handle it differently inside  `set_weights_and_activate_adapters`

UmerHA commented:
> Can we just a LoRA from the guide I mentioned above and show a simple comparative study given we already know that it doesn't have a text encoder?

Ah okay, I misunderstood you. Sure. I've now used the LoRas from the guide and removed the text-encoder scale, like so:
```py
prompt = "toy_face of a hacker with a hoodie, pixel art"

adapter_weight_scales = { "unet": { "down": 1, "mid": 0, "up": 0} }
pipe.set_adapters("pixel", adapter_weight_scales)
...
adapter_weight_scales = { "unet": { "down": 0, "mid": 1, "up": 0} }
pipe.set_adapters("pixel", adapter_weight_scales)
...
adapter_weight_scales = { "unet": { "down": 0, "mid": 0, "up": 1} }
pipe.set_adapters("pixel", adapter_weight_scales)
...
adapter_weight_scales_toy = 0.5
adapter_weight_scales_pixel = { ... }
pipe.set_adapters(["toy", "pixel"], [adapter_weight_scales_toy, adapter_weight_scales_pixel])
```

The resulting image are the same is in https://github.com/huggingface/diffusers/pull/7352#issuecomment-2013890540.

UmerHA commented:
> I left a few more feedback, basically, I think `set_weights_and_activate_adapters` should still be able to handle the `weights` argument in the previous format (a list of scalar or float); this way we only need to expand the `weights` when it is a `dict` and handle it differently inside `set_weights_and_activate_adapters`

All done

yiyixuxu reviewed (approved):
looks good to me now!
@sayakpaul feel free to merge once you're happy with it

yiyixuxu commented on src/diffusers/loaders/lora.py:
```diff
@@ -1033,17 +1039,78 @@ def enable_lora_for_text_encoder(self, text_encoder: Optional["PreTrainedModel"]
+                        "Lora weight dict contains text_encoder_2 weights but will be ignored because pipeline does not have text_encoder_2."
+                    )
+
+                # warn if adapter doesn't have parts specified by adapter_weights
```
can we move these warnings into `maybe_expand_lora_scales`?

UmerHA replied:
see below

yiyixuxu commented on src/diffusers/loaders/lora.py:
```diff
@@ -1033,17 +1039,78 @@ def enable_lora_for_text_encoder(self, text_encoder: Optional["PreTrainedModel"]
+                f"Length of adapter names {len(adapter_names)} is not equal to the length of the weights {len(adapter_weights)}"
+            )
+
+        def warn_if_adapter_misses_part(adapter_name, part_name):
```
can we move this function to maybe_expand_lora_scales?

UmerHA replied:
Have moved it & the warnings. Because they're not unet-specific, I've renamed the file from `unet_loader_utils` to `lora_loader_utils.py`

yiyixuxu commented on src/diffusers/loaders/lora.py:
```diff
@@ -1033,17 +1040,67 @@ def enable_lora_for_text_encoder(self, text_encoder: Optional["PreTrainedModel"]
+
+        # Decompose weights into weights for unet, text_encoder and text_encoder_2
+        unet_lora_weights, text_encoder_lora_weights, text_encoder_2_lora_weights = [], [], []
+
```
```suggestion
adapters = self.get_list_adapters()
invert_list_adapters = {adapter_name: model for model, adapter_name in adapters.items()}
```

yiyixuxu commented on src/diffusers/loaders/lora.py:
```diff
@@ -1033,17 +1040,67 @@ def enable_lora_for_text_encoder(self, text_encoder: Optional["PreTrainedModel"]
+                    )
+
+                # warn if adapter doesn't have parts specified by adapter_weights
+                warn_if_adapter_and_scales_mismatch(
+                    adapter_name,
+                    self.get_list_adapters(),
+                    unet_lora_weight,
+                    text_encoder_lora_weight,
+                    text_encoder_2_lora_weight,
+                )
```
```suggestion
if unet_lora_weight is not None and "unet" not in invert_list_adapters[adapter_name]:
    logger.warning(" ...")
if text_encoder_lora_weight is not None and "text_encoder" not in invert_list_adapters[adapter_name]:
   logger.warning(" ...")
...
```

yiyixuxu commented on src/diffusers/loaders/lora_loader_utils.py:
```diff
@@ -0,0 +1,175 @@
+    return {translate_into_actual_layer_name(name): weight for name, weight in scales.items()}
+
+
+def warn_if_adapter_and_scales_mismatch(
+    adapter_name, adapter_list, unet_lora_weight, text_encoder_lora_weight, text_encoder_2_lora_weight
+):
+    """Warn if the scales dict for an adapter contain parts (unet, text_encoder, text_encoder_2) for which the adapter doesn't have weights."""
+    adapter_parts = [
+        part for part, adapters in adapter_list.items() for adapter in adapters if adapter == adapter_name
+    ]
+
+    for part_name, part_weight_scales in zip(
+        ["unet", "text_encoder", "text_encoder_2"],
+        [unet_lora_weight, text_encoder_lora_weight, text_encoder_2_lora_weight],
+    ):
+        if part_weight_scales is None:
+            continue
+
+        if part_name not in adapter_parts:
+            logger.warning(
+                f"Lora weight dict for adapter '{adapter_name}' contains {part_name}, but this will be ignored because {adapter_name} does not contain weights for {part_name}. Valid parts for {adapter_name} are: {adapter_parts}."
+            )
```
```suggestion
```

yiyixuxu reviewed (commented):
ohh sorry I misunderstood, I thought the warnings are specific to unet

would something like this work?

UmerHA commented:
> would something like this work?

With a small adjustment, yes. `invert_list_adapters` needs to be created a bit differently, because `self.get_list_adapters()` is not `Dict[str, str]` but `Dict[str, List[str]]`.

This works:
```py
list_adapters = self.get_list_adapters()  # eg {"unet": ["adapter1", "adapter2"], "text_encoder": ["adapter2"]}
all_adapters = {
    adapter for adapters in list_adapters.values() for adapter in adapters
}  # eg ["adapter1", "adapter2"]
invert_list_adapters = {
    adapter: [part for part, adapters in list_adapters.items() if adapter in adapters]
    for adapter in all_adapters
}  # eg {"adapter1": ["unet"], "adapter2": ["unet", "text_encoder"]}
```

I have adjusted to code accordingly, and renamed `lora_loaders_util.py` back to `unet_loaders_util.py`

UmerHA commented:
@sayakpaul gentle ping, as this has been stale for a week

sayakpaul commented:
Will do a thorough review tomorrow first thing. Thanks for your patience.

sayakpaul commented on src/diffusers/loaders/unet_loader_utils.py:
```diff
@@ -0,0 +1,154 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import copy
```
Also, remind me why are these in `unet_loader_utils.py` and not in `lora_loader_utils.py`? I assume some of these stuff also get applied to the text encoder LoRAs no?

UmerHA replied:
No, they're only applied to the unet parts. We scale the text encoder as a whole. The user can only pass a single number and it's not expanded. Therefore, these functions are not needed for it.

Now that I think of it, it's maybe not reasonable to assume the user won't want to scale the text encoder more granularly.

But I'd say we ship this PR and if there's demand, I can add support for scaling the text encoder more granularly.

WDYT?

sayakpaul replied:
100 percent agreed. Thank you for walking me through this.

sayakpaul reviewed (commented):
Left a last couple of comments. Looking really nice. Going to run some low tests to ensure we're not breaking anything here. 

Really amazing work!

sayakpaul commented:
LoRA SD slow tests are passing which should be more than enough to confirm the PR isn't backward breaking. Let's address the last comments and ship this beast :shipit:

UmerHA commented on docs/source/en/tutorials/using_peft_for_inference.md:
````diff
@@ -133,6 +133,62 @@ image
+image
+```
+
+![block-lora-text-and-down](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/peft_integration/diffusers_peft_lora_inference_block_down.png)
````
I've uploaded them. The PR is here: https://huggingface.co/datasets/huggingface/documentation-images/discussions/310.

As I'm doing this for the firs time: Can you quickly verify it worked? Otherwise, I can also send the images over via mail

UmerHA commented:
@sayakpaul Everything is addressed! Iiuc, you now have to accept the PR for the uploaded images (https://huggingface.co/datasets/huggingface/documentation-images/discussions/310) and then we're done :shipit::shipit::shipit:

sayakpaul commented:
Already merged.

asomoza commented:
really nice, when I have the time I'll play a lot with this and maybe post my results so people can understand better what can they do with this level of control over the LoRAs

sayakpaul commented:
@UmerHA could you push an empty commit to start the CI? And please ping me once you do.

sayakpaul commented:
Thanks @UmerHA!

Will wait for the CI to run and will then ship. 

To help promote this amazing feature, I welcome you to open a detailed thread here: https://github.com/huggingface/diffusers/discussions. 

Additionally, if you're on HF Discord, please let me know your username so that we can give you a shoutout. If you're not, you can join via this link: https://hf.co/join.discord.

UmerHA commented:
> Thanks @UmerHA!

❤️

> To help promote this amazing feature, I welcome you to open a detailed thread here: https://github.com/huggingface/diffusers/discussions.

Will do this evening/tomorrow ✅

> Additionally, if you're on HF Discord, please let me know your username so that we can give you a shoutout. If you're not, you can join via this link: https://hf.co/join.discord.

Thanks :) My discord is `umerha`. But my primary channel is Twitter where I am [@UmerHAdil](https://x.com/UmerHAdil). If possible, I'd highly appreciate a shootout over there 🙏🏽

asomoza commented:
@UmerHA I'll leave this here because I can't do a PR right now for quick fix.

Just updated my main and ran the test I did before, there a small typo `uent` that throws a warning:

```bash
Lora weight dict for adapter 'lora' contains uent, but this will be ignored because lora does not contain weights for uent. Valid parts for lora are: ['text_encoder', 'text_encoder_2', 'unet'].
```

UmerHA commented:
> there a small typo `uent` that throws a warning:

@asomoza thanks! have made a fix PR [here](https://github.com/huggingface/diffusers/pull/7523)

UmerHA commented:
> > To help promote this amazing feature, I welcome you to open a detailed thread here: https://github.com/huggingface/diffusers/discussions.
> 
> Will do this evening/tomorrow ✅

Done ✅
